### PR TITLE
[1404] Added basic validation message

### PR DESCRIPTION
**Pull request by defong**
### Context

Implement validation messages for publishing a course.

### Changes proposed in this pull request

- Parse and display errors from backend
- Deep link to manage-courses-ui from the errors

### Guidance to review

Screenshot 2019-05-15 at 13 06 43

### CODE REVIEW

File: app/controllers/courses_controller.rb [hunk 1]
```diff
@@ -44,9 +44,20 @@ def withdraw; end
...
+        {
+          key: error[:title].last(error[:title].length - 'Invalid latest_enrichment__'.length),
+          value: error[:detail]
+        }
+      end
+
+      redirect_to description_provider_course_path(@provider.provider_code, @course.course_code)
+    else
+      flash[:success] = 'Your changes have been published'
+      redirect_to description_provider_course_path(@provider.provider_code, @course.course_code)
```
Comment (dankmitchell) [inline app/controllers/courses_controller.rb]: As this is repeated in both conditions we can move it out

File: app/controllers/courses_controller.rb [hunk 2]
```diff
@@ -44,9 +44,20 @@ def withdraw; end
   def delete; end
 
   def publish
-    @course.publish(provider_code: @provider.provider_code)
-    flash[:success] = 'Your changes have been published'
-    redirect_to description_provider_course_path(@provider.provider_code, @course.course_code)
+    errors = @course.publish(provider_code: @provider.provider_code).errors
+    if errors.present?
+      flash[:publish_errors] =  errors.map do |error|
```
Comment (dankmitchell) [inline app/controllers/courses_controller.rb]: Extra space here

File: app/views/layouts/_flash.html.erb [hunk 3]
```diff
@@ -1,9 +1,13 @@
 <div class="grid-row">
   <div class="column-two-thirds">
-    <% flash.each do |key, value| %>
-      <div class="govuk-<%= key %>-summary" role="alert" aria-labelledby="<%= key %>-summary-heading" tabindex="-1" data-module="error-summary">
-        <h3 class="govuk-<%= key %>-summary__title" id="<%= key %>-summary-heading"><%= value %></h3>
-      </div>
+    <% if flash[:publish_errors].any?%>
```
Comment (dankmitchell) [inline app/views/layouts/_flash.html.erb]: Missing space here

File: app/views/layouts/_flash.html.erb [hunk 4]
```diff
@@ -1,9 +1,13 @@
...
+      <%= render partial: 'layouts/publish_errors' %>
```
Comment (dankmitchell) [inline app/views/layouts/_flash.html.erb]: We could make this generic as I'm sure we'll re-use it and call it `error_summary`

File: app/views/layouts/_publish_errors.html.erb [hunk 5]
```diff
@@ -0,0 +1,18 @@
+<div class="govuk-error-summary" aria-labelledby="error-summary-title" role="alert" tabindex="-1" data-module="error-summary">
```
Comment (dankmitchell) [inline app/views/layouts/_publish_errors.html.erb]: `_ error_summary.html.erb`?

**Comment by defong** [comment]
[Frontend course publishing validations](https://trello.com/c/AC5lOEqg/1404-frontend-course-publishing-validations)

**Comment by tvararu** [comment]
Have updated this PR with tests and it's in a reviewable state.

Before it's merged, I will clean up the commit history.

File: app/controllers/courses_controller.rb [hunk 6]
```diff
@@ -50,8 +53,18 @@ def withdraw; end
...
+    errors = @course.publish(provider_code: @provider.provider_code).errors
+    if errors.present?
+      flash[:error_summary] = errors.map { |error|
+        [
+          error[:title].last(error[:title].length - 'Invalid latest_enrichment__'.length),
+          error[:detail]
+        ]
+      }.to_h
+    else
+      flash[:success] = 'Your changes have been published'
```
Comment (defong) [inline app/controllers/courses_controller.rb]: image

?

File: app/helpers/course_helper.rb [hunk 7]
```diff
@@ -1,22 +1,39 @@
...
+                  provider_code: @provider.provider_code,
+                  course: @course,
+                  field: field
+                )
   end
 
-  def course_summary_value(value, data)
+  def course_summary_label(key, field)
+    error = @errors.present? && @errors[field]
+    if error.present?
```
[the remark this answers is not in the document]
Comment (misaka) [reply app/helpers/course_helper.rb]: ^^^ minor issue, happy to pass this without

[the remark this answers is not in the document]
Comment (tvararu) [reply app/helpers/course_helper.rb]: I like it

File: app/controllers/courses_controller.rb (hunk 6 above)
Comment (tvararu) [reply app/controllers/courses_controller.rb]: Done...

**Pull request merged**